### PR TITLE
Service Worker Resilience: Retries of caching attempts

### PR DESCRIPTION
**Pull request by manuelroemer**
Still WIP, but I want to gather some initial feedback of whether this approach is right. Feel free to review!

# Open Tasks / Questions

- [x] Testing of the behavior.
- [x] Q: When should we retry? Like it is currently (during _every_ caching attempt) or only in selected cases? 
- [x] Q: Are the retry "values", i.e. num of attempts, duration etc. good enough? (I think they might have to be increased).

### CODE REVIEW

**Comment by FlorianRappl** [comment]
> When should we retry? Like it is currently (during every caching attempt) or only in selected cases?

Sounds fine to me. Everything we want to cache is something we should really try to get loaded. So yes.

> Are the retry "values", i.e. num of attempts, duration etc. good enough? (I think they might have to be increased).

Personally, I think they are fine. It should be between 3 and 5 entries and the total max. length may not exceed 40 seconds if you ask me. But the chosen values are fine.

Regarding the testing of the behavior; good question. I need to think about it. Honestly, we need to have an E2E testing strategy here (automated). I think we need to ask for support here as presumably there is already the QA team active.

File: packages/shell/esm-app-shell/src/service-worker/caching.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
-import { ImportMap } from "@openmrs/esm-globals";
-import { absoluteWbManifestUrls, indexUrl, omrsCacheName } from "./constants";
+import { ImportMap } from "@openmrs/esm-framework";
+import { retry } from "@openmrs/esm-utils";
```
Comment (manuelroemer) [inline packages/shell/esm-app-shell/src/service-worker/caching.ts:2]: I faced some weird issue during debugging just now where the SW could no longer be loaded because apparently the JS file was invalid. This happens specifically when importing `retry` from `esm-framework`. It works just fine with the `esm-utils` import here. 

Any idea what might be causing this?

Comment (FlorianRappl) [reply packages/shell/esm-app-shell/src/service-worker/caching.ts:2]: Yes, don't import framework. Its very heavy and the service worker should - if any - just rely on `esm-offline`. `esm-framework` also contains `esm-globals`, which works against the DOM. In the service worker you don't have access to the DOM, so this will not work.

Comment (FlorianRappl) [reply packages/shell/esm-app-shell/src/service-worker/caching.ts:2]: Note that the import of `ImportMap` does not have any impact as this is a type, so TS will erase the actual import (and you will not face the DOM access issue there).

Comment (manuelroemer) [reply packages/shell/esm-app-shell/src/service-worker/caching.ts:2]: That makes a lot of sense suddenly, thanks!

Comment (manuelroemer) [reply packages/shell/esm-app-shell/src/service-worker/caching.ts:2]: I went over the SW again and replaced one lingering esm-framework import with esm-global.

File: packages/framework/esm-api/src/types/visit-resource.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { OpenmrsResource } from "./openmrs-resource";
 
 export interface NewVisitPayload {
+  uuid?: string;
```
Comment (manuelroemer) [inline packages/framework/esm-api/src/types/visit-resource.ts:4]: Smuggling this one in so that I don't have to do another PR for this small change.

Comment (FlorianRappl) [reply packages/framework/esm-api/src/types/visit-resource.ts:4]: ;) Works!

File: packages/shell/esm-app-shell/src/service-worker/caching.ts (hunk 1 above)
Comment (FlorianRappl) [reply packages/shell/esm-app-shell/src/service-worker/caching.ts:2]: Great!

**Review by FlorianRappl — APPROVED** [review]
Great! :beers:
